app/ide-desktop/lib/profiling/src/profiling.ts

```diff
@@ -0,0 +1,322 @@
+    /// an interval. The label encoded the log level, name and event type.
+    private startIntervalLabel(interval_name: string): string {
+        return this.createLabel(interval_name, 'start')
+    }
+
+    /// Return a string that encodes the given log level and name for a mark that indicates the end of
+    /// an interval.
+    private endIntervalLabel(interval_name: string): string {
+        return this.createLabel(interval_name, 'end')
+    }
+
+    /// Return a string that encodes the given log level and name for a measurement.
+    private measureIntervalLabel(interval_name: string): string {
+        return this.createLabel(interval_name, 'measure')
+    }
+
+    /// Start the profiling of the named interval.
+    start(interval_name: string): IntervalHandle {
+        if (this.isActive()) {
+            performance.mark(this.startIntervalLabel(interval_name))
+        }
+        return new IntervalHandle(interval_name, this)
+    }
+
+    /// End the profiling of the named interval.
+    end(interval_name: string) {
+        let start_label = this.startIntervalLabel(interval_name)
+        let end_label = this.endIntervalLabel(interval_name)
+        let measurement_label = this.measureIntervalLabel(interval_name)
+
+        if (this.isActive()) {
+            performance.mark(end_label)
+            performance.measure(measurement_label, start_label, end_label)
+        }
+    }
+
+    /// Profile the execution of the given callable.
+    measure(interval_name: string, closure: CallableFunction) {
+        this.start(interval_name)
+        const ret = closure()
+        this.end(interval_name)
+        return ret
+    }
+}
+
+// =========================
+// === Encoding/Decoding ===
+// =========================
+
+/// Delimiter used to to encode information in the `PerformanceEntry` name.
+const MESSAGE_DELIMITER = '//'
+
+/**
+ * Encodes the given information in a string. This is done by separating the information by the `MESSAGE_DELIMITER`.
+ * Example output: "TASK//SomeWork//start"
+ */
+function encodeLabel(log_level: Profiler, inter_val_name: string, event: string): string {
+    return [log_level.log_level, inter_val_name, event].join(MESSAGE_DELIMITER)
+}
+
+function parseLogLevel(log_level_name: string): Profiler | undefined {
+    for (const profiler of profiler_registry) {
+        if (profiler.isLogLevelName(log_level_name)) {
+            return profiler
+        }
+    }
+    return undefined
+}
+
+/**
+ * Decodes the given information in a string that was encoded with `encodeLabel`.
+ */
+function decodeLabel(label: string): [Profiler, string] | undefined {
+    let parts = label.split(MESSAGE_DELIMITER)
+    try {
+        let [log_level_name, interval_name] = [parts[0], parts[1]]
+        let profiler = parseLogLevel(log_level_name)
+        return [profiler, interval_name]
+    } catch (e) {
+        return undefined
```

Comment:
Never use undefined as a value. The connotation of undefined is "this was never defined by the programmer". Use `null` instead. It is just bad practice to use undefined – this should be used only as a missing value (never defined).

Reply:
Changed to `null` in all relevant places.